python/wikibot/wikibot.py

```diff
@@ -0,0 +1,183 @@
+        continue
+      wd_claims = wd_item.get().get('claims')
+      fact_record = rs.parse(record)
+      if rs[item] != fact_record[self.n_item]:
+        print "Inconsistent key & item input: ", item, fact_record[self.n_item]
+        status_record = rs.frame({
+          self.n_item: fact_record[self.n_item],
+          self.n_facts: fact_record[self.n_facts],
+          self.n_status: rs.frame({self.n_skipped: self.n_incons_input})
+        })
+        self.status_file.write(item, status_record.data(binary=True))
+        continue # read next record in the file
+      provenance = fact_record[self.n_provenance]
+      # Process facts / claims
+      facts = fact_record[self.n_facts]
+      for prop, val in facts:
+        prop_str = str(prop)
+        if prop_str in wd_claims:
+          print item, "already has property", prop
+          status_record = rs.frame({
+            self.n_item: fact_record[self.n_item],
+            self.n_facts: rs.frame({prop: val}),
+            self.n_status: rs.frame({self.n_skipped: self.n_has_prop})
+          })
+          self.status_file.write(item, status_record.data(binary=True))
+          continue
+        if self.everHadProp(wd_item, prop_str):
+          print item, "previously had property", prop
+          status_record = rs.frame({
+            self.n_item: fact_record[self.n_item],
+            self.n_facts: rs.frame({prop: val}),
+            self.n_status: rs.frame({self.n_skipped: self.n_had_prop})
+          })
+          self.status_file.write(item, status_record.data(binary=True))
+          continue
+        claim = pywikibot.Claim(self.repo, prop_str)
+        if claim.type == "time":
+          date = sling.Date(val) #parse date from record
+          precision = precision_map[date.precision] # sling to wikidata
+          target = pywikibot.WbTime(year=date.year, precision=precision)
+        elif claim.type == 'wikibase-item':
+          target = pywikibot.ItemPage(self.repo, val)
+        else:
+          # TODO add location and possibly other types
+          print "Error: Unknown claim type", claim.type
+          continue
+        claim.setTarget(target)
+        cat_str = str(provenance[self.n_category])
+        summary = provenance[self.n_method] + " " + cat_str
+        wd_item.addClaim(claim, summary=summary)
+        claim.addSources(self.getSources(cat_str))
+        rev_id = wd_item.latest_revision_id
+        status_record = rs.frame({
+          self.n_item: fact_record[self.n_item],
+          self.n_facts: rs.frame({prop: val}),
+          self.n_status: rs.frame({self.n_stored: str(rev_id)})
+        })
+        self.status_file.write(item, status_record.data(binary=True))
+
+      updated += 1
+      print item
+    print "Last record.", updated, "records updated."
+    self.status_file.close()
+    self.record_file.close()
+
+  def run(self):
+    self.storeRecords(self.record_file, batch_size=2)
+
+if __name__ == '__main__':
+  sfb = StoreFactsBot(test=True)
```

Comment:
Should test be True here?

Reply:
Moved this boolean out so it is a cmd-line flag instead. HTH.